### PR TITLE
compiler: add ability to generate .abi.json file

### PR DESCRIPTION
**Pull request by AnnaShaleva**
A part of integration with NEO Blockchain Toolkit (see #902). To be
able to deploy smart-contract compiled with neo-go compiler via NEO
Express, we have to generate additional .abi.json file. This file
contains the following information:
 - hash of the compiled contract
 - smart-contract metadata (title, description, version, author,
email, has-storage, has-dynamic-invoke, is-payable)
 - smart-contract entry point
 - functions
 - events

However, this .abi.json file is slightly different from the one,
described in manifest.go, so we have to add auxilaury stractures for
json marshalling. The .abi.json format used by NEO-Express is described
[here](https://github.com/neo-project/neo-devpack-dotnet/blob/master/src/Neo.Compiler.MSIL/FuncExport.cs#L66).

### CODE REVIEW

File: pkg/compiler/compiler.go [hunk 1]
```diff
@@ -26,6 +27,12 @@ type Options struct {
 
 	// The name of the output for debug info.
 	DebugInfo string
+
+	// The name of the output for application binary interface info.
+	ABIInfo string
+
+	// Contract metadata.
+	ContractDetails *request.ContractDetails
```
Comment (roman-khimov) [inline pkg/compiler/compiler.go:35]: Please move `ContractDetails` out of `rpc/request` to `smartcontract`, importing anything RPC-related into the compiler is just weird.

File: pkg/compiler/debug.go [hunk 2]
```diff
@@ -260,3 +298,40 @@ func parsePairJSON(data []byte, sep string) (string, string, error) {
 	}
 	return ss[0], ss[1], nil
 }
+
+func (di *DebugInfo) convertToABI(contract []byte, cd *request.ContractDetails) ABI {
```
Comment (roman-khimov) [inline pkg/compiler/debug.go:302]: Can we add some test for it?

**Comment by AnnaShaleva** [comment]
Also as part of #902, for the current moment we have difference in abi.json parameters/variables type format which leads to `Invalid variable attributes` error in NEO Express. The difference is in serialization of non-standard types: 
 - c# implementation has `Unknown: {type}` (see [for details](https://github.com/neo-project/neo-devpack-dotnet/blob/0c68a4f589dab8fb9c2352259acd6e28523441df/src/Neo.Compiler.MSIL/FuncExport.cs#L63))
 - go implementation has `Any`, which can not be properly parsed by NEO debugger.

However, that's not critical error.

File: pkg/compiler/compiler.go [hunk 3]
```diff
@@ -11,6 +11,8 @@ import (
 	"path/filepath"
 	"strings"
 
+	"github.com/nspcc-dev/neo-go/pkg/smartcontract"
+
```
Comment (roman-khimov) [inline pkg/compiler/compiler.go:15]: Blank line.

File: pkg/compiler/debug.go [hunk 4]
```diff
@@ -8,6 +8,11 @@ import (
 	"go/types"
 	"strconv"
 	"strings"
+
+	"github.com/nspcc-dev/neo-go/pkg/smartcontract"
+
```
Comment (roman-khimov) [inline pkg/compiler/debug.go:13]: Blank line.

File: pkg/compiler/debug_test.go [hunk 5]
```diff
@@ -3,6 +3,9 @@ package compiler
 import (
 	"testing"
 
+	"github.com/nspcc-dev/neo-go/pkg/crypto/hash"
+	"github.com/nspcc-dev/neo-go/pkg/smartcontract"
+
```
Comment (roman-khimov) [inline pkg/compiler/debug_test.go:8]: Blank line.

File: pkg/compiler/debug.go [hunk 6]
```diff
@@ -111,13 +111,15 @@ func (c *codegen) registerDebugVariable(name string, expr ast.Expr) {
 
 func (c *codegen) methodInfoFromScope(name string, scope *funcScope) *MethodDebugInfo {
 	ps := scope.decl.Type.Params
-	params := make([]DebugParam, 0, ps.NumFields())
-	for i := range params {
+	params := make([]DebugParam, ps.NumFields())
+	count := 0
```
Comment (roman-khimov) [inline pkg/compiler/debug.go:154]: I think `append()` style would still look a bit better here (not needing any `count`, for example), to fix the original problem you only need to change the range iterator.

**Pull request merged**